### PR TITLE
Feature/getjoke

### PR DESCRIPTION
**Pull request by fandomario**
https://forums.pythonic.guru/t/topic/8799

### CODE REVIEW

**Comment by PureDreamer** [comment]
Please change the pr to PythonFreeCourse:develop!

File: tests/test_getjoke.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from app.routers import getjoke
+
+
+def test_getjoke(client):
+    resp = client.get('/getjoke')
+    assert resp.ok
+    assert resp.json
```
Comment (PureDreamer) [inline tests/test_getjoke.py]: I would do a mock to see if the response is really a joke.

File: app/templates/base.html [hunk 2]
```diff
@@ -45,6 +45,16 @@
...
                     </li>
+                    <li class="nav-item">
+                        <button onclick="getDailyJoke()" class="btn btn-link" style='text-decoration:none;'>get a joke</button>
+                        <script>
+                        function getDailyJoke() {
+                          fetch('/getjoke')
+                            .then(response => response.json())
+                            .then(data => alert(data.value.joke));
+                        }
+                        </script>
```
Comment (PureDreamer) [inline app/templates/base.html]: It would be better if you could take it out to another function(a JavaScript one).

File: app/routers/getjoke.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from fastapi import APIRouter
+import requests
+
+
+router = APIRouter()
+
+
+@router.get("/getjoke")
+def getjoke():
+    return requests.get(f'http://api.icndb.com/jokes/random').json()
```
Comment (PureDreamer) [inline app/routers/getjoke.py]: this should be on internal as it is an internal function, that is needed to be called from a route.

File: app/templates/base.html [hunk 4]
```diff
@@ -45,6 +45,10 @@
                     <li class="nav-item">
                         <a class="nav-link" href="/search">Search</a> 
                     </li>
+                    <li class="nav-item">
+                        <button onclick="getDailyJoke()" class="btn btn-link" style='text-decoration:none;'>Get a Joke</button>
```
Comment (yammesicka) [inline app/templates/base.html]: Prefer using `addEventHandler`

File: app/routers/getjoke.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import requests
+from fastapi import APIRouter
+
+
+router = APIRouter()
+
+
+@router.get("/getjoke")
```
Comment (yammesicka) [inline app/routers/getjoke.py]: Just `/joke`, the GET is the HTTP Verb itself

File: app/static/joke.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+function getDailyJoke() {
+  fetch('/getjoke')
+    .then(response => response.json())
+    .then(data => alert(data.value.joke));
```
Comment (yammesicka) [inline app/static/joke.js]: You can do much better :P

File: app/routers/getjoke.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+def getjoke():
+    return requests.get('http://api.icndb.com/jokes/random').json()
```
Comment (yammesicka) [inline app/routers/getjoke.py]: Can you rewrite it to be on the frontend side?
Every HTTP call from the server side might majorly degrade our performance.

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/routers/getjoke.py (hunk 5 above)
Comment (fandomario) [reply app/routers/getjoke.py]: done.

File: app/routers/getjoke.py (hunk 7 above)
Comment (fandomario) [reply app/routers/getjoke.py]: yes.

File: app/internal/jokes.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+from app.database.models import Joke
+
+from sqlalchemy.orm import Session
+from sqlalchemy.sql.expression import func
+
+
+def have_a_joke(session: Session):
+    joke = session.query(Joke).order_by(func.random()).first()
+    return joke
```
Comment (Gonzom) [inline app/internal/jokes.py]: Change name to "get_a_joke".

File: app/templates/base.html [hunk 9]
```diff
@@ -45,6 +45,10 @@
                     <li class="nav-item">
                         <a class="nav-link" href="/search">Search</a> 
                     </li>
+                    <li class="nav-item">
+                        <button onclick="makejoke()" class="btn btn-link" style='text-decoration:none;'>Make me Laugh</button>
```
Comment (Gonzom) [inline app/templates/base.html]: Move the style to the CSS file.

File: app/routers/joke.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+from sqlalchemy.orm import Session
+from app.database.database import get_db
+
+
+router = APIRouter()
+
+@router.get("/joke")
+async def joke(request: Request, db: Session = Depends(get_db)):
+    joke = jokes.have_a_joke(db)
+    return joke
```
Comment (Gonzom) [inline app/routers/joke.py]: You can return the object without needing to assign it.

File: app/templates/base.html (hunk 4 above)
Comment (yammesicka) [reply app/templates/base.html]: (This one is still relevant)

File: app/templates/base.html [hunk 11]
```diff
@@ -45,6 +45,10 @@
                     <li class="nav-item">
                         <a class="nav-link" href="/search">Search</a> 
                     </li>
+                    <li class="nav-item">
+                        <button onclick="getDailyJoke()" class="btn btn-link" style='text-decoration:none;'>Get a Joke</button>
+                        <script src="../static/joke.js"></script>
```
Comment (yammesicka) [inline app/templates/base.html]: Prefer using url_to

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/templates/base.html (hunk 2 above)
Comment (fandomario) [reply app/templates/base.html]: I did

File: app/routers/getjoke.py (hunk 3 above)
Comment (fandomario) [reply app/routers/getjoke.py]: Done

File: app/routers/getjoke.py (hunk 7 above)
Comment (fandomario) [reply app/routers/getjoke.py]: I did

File: tests/test_getjoke.py (hunk 1 above)
Comment (fandomario) [reply tests/test_getjoke.py]: I have made some new tests to check also that

File: app/templates/base.html (hunk 11 above)
Comment (fandomario) [reply app/templates/base.html]: Done

File: app/templates/base.html (hunk 4 above)
Comment (fandomario) [reply app/templates/base.html]: I am trying to do so.
I didn't succeed yet.

File: app/templates/base.html (hunk 9 above)
Comment (yammesicka) [reply app/templates/base.html]: Also the JavaScript, please :) (use `addEventListener`)

File: app/internal/jokes.py [hunk 12]
```diff
@@ -0,0 +1,9 @@
+from app.database.models import Joke
+
+from sqlalchemy.orm import Session
+from sqlalchemy.sql.expression import func
+
+
+def get_a_joke(session: Session):
+    joke = session.query(Joke).order_by(func.random()).first()
```
Comment (yammesicka) [inline app/internal/jokes.py]: You can directly return the value

**Review by yammesicka — COMMENTED** [review]
You've closed comments that you haven't fixed yet.
Please don't to it.

Go back and fix the relevant unfixed comments.

File: app/templates/base.html (hunk 4 above)
Comment (fandomario) [reply app/templates/base.html]: I have changed everything according to your requests :)
I  added an event listener

File: app/static/joke.js (hunk 6 above)
Comment (fandomario) [reply app/static/joke.js]: I didn't really understand this one :)
I have change the logic with an event listener. I really like it popping. If it's ok.

File: app/templates/base.html (hunk 9 above)
Comment (fandomario) [reply app/templates/base.html]: I  added an event listener and cancel the need in css file

File: app/internal/jokes.py (hunk 12 above)
Comment (fandomario) [reply app/internal/jokes.py]: I have fixed that.

File: app/static/joke.js (hunk 6 above)
Comment (yammesicka) [reply app/static/joke.js]: If you like the popup, please use [Bootstrap Models](https://getbootstrap.com/docs/5.0/components/modal/) instead - and not JS alert

File: app/static/joke.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+
+function addEventsAfterPageLoaded() {
+  const element = document.getElementById("ajoke");
+  element.addEventListener("click", makejoke, false);
+  };
+
+
+document.addEventListener("DOMContentLoaded", addEventsAfterPageLoaded);
```
Comment (yammesicka) [inline app/static/joke.js:14]: Awesome!

File: app/static/joke.js [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+  fetch('/joke')
+    .then(response => response.json())
+    .then(data => alert(data.text));
+}
+
+
+function addEventsAfterPageLoaded() {
+  const element = document.getElementById("ajoke");
+  element.addEventListener("click", makejoke, false);
+  };
```
Comment (yammesicka) [inline app/static/joke.js]: This shouldn't be indented.
Also, please remove the `;`

File: app/templates/base.html [hunk 15]
```diff
@@ -51,6 +51,9 @@
                         <li class="nav-item">
                             <a class="nav-link" href="{{ url_for('search') }}">Search</a>
                         </li>
+                        <li class="nav-item">
+                            <button onclick="makejoke()" class="btn btn-link">Make me Laugh</button>
```
Comment (yammesicka) [inline app/templates/base.html]: Remove the `onclick`

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/static/joke.js (hunk 6 above)
Comment (PureDreamer) [reply app/static/joke.js]: You can also use the JS library that I used for the text editor.
It's called sweetalert2, its in the text_editor.js

File: app/templates/base.html (hunk 15 above)
Comment (fandomario) [reply app/templates/base.html]: Done

File: app/static/joke.js (hunk 14 above)
Comment (fandomario) [reply app/static/joke.js]: Done

File: app/static/joke.js (hunk 13 above)
Comment (fandomario) [reply app/static/joke.js:14]: I received help for that one.
Thanks :)

File: app/internal/jokes.py (hunk 12 above)
Comment (fandomario) [reply app/internal/jokes.py]: Done

File: app/internal/load_jokes.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    return jokes_list
+
+
+def add_jokes_to_db(session: Session) -> None:
+    """This function reads the jokes and inserts them into the db"""
+    all_jokes = get_jokes_from_json()
+    jokes_objects = [
+        Joke(text=joke['text'])
+        for joke in all_jokes
+        ]
```
Comment (yammesicka) [inline app/internal/load_jokes.py]: Can be in a single line

File: app/static/joke.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
+function makejoke() {
+  fetch('/joke')
+    .then(response => response.json())
+    .then(data => alert(data.text));
```
Comment (yammesicka) [inline app/static/joke.js]: Almost perfect!
Now use [SweetAlert2](https://sweetalert2.github.io/) to replace the `alert`. It should be easy and would look much better.

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/load_jokes.py (hunk 16 above)
Comment (fandomario) [reply app/internal/load_jokes.py]: Done

File: app/static/joke.js (hunk 17 above)
Comment (fandomario) [reply app/static/joke.js]: Thanks :)
Done

File: app/static/joke.js (hunk 6 above)
Comment (fandomario) [reply app/static/joke.js]: I used SweetAlert2 as you suggested

File: app/static/joke.js (hunk 17 above)
Comment (yammesicka) [reply app/static/joke.js]: Still here

File: app/internal/load_jokes.py (hunk 16 above)
Comment (yammesicka) [reply app/internal/load_jokes.py]: ?

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: app/internal/load_jokes.py (hunk 16 above)
Comment (fandomario) [reply app/internal/load_jokes.py]: Something is wrong. I have changed everything you asked for and in my computer it is deferent from here.
It happened the last time too.
I am going to check why it is happening,
Sorry

Comment (fandomario) [reply app/internal/load_jokes.py]: Maybe it is because I pushed once 2 pr on the same branch

Comment (Gonzom) [inline app/internal/load_jokes.py]: This seems to be the same code as json_data_loader.py, why not use it instead?

File: app/main.py [hunk 18]
```diff
@@ -42,10 +42,11 @@ def create_tables(engine, psql_environment):
 from app.routers import (  # noqa: E402
     about_us, agenda, calendar, categories, celebrity, credits,
     currency, dayview, email, event, export, four_o_four, friendview,
-    google_connect, invitation, login, logout, profile,
+    google_connect, invitation, joke, login, logout, profile,
     register, search, telegram, user, weekview, whatsapp,
 )
 
+load_jokes.load_daily_jokes(next(get_db()))
 json_data_loader.load_to_database(next(get_db()))
```
Comment (Gonzom) [inline app/main.py]: Why not use json_data_loader instead like the others?

File: app/internal/load_jokes.py (hunk 16 above)
Comment (fandomario) [reply app/internal/load_jokes.py]: I fixed that. It was exactly what I thought. The files from my computer didn't push here.
Any way. Now it is really done.
Sorry for your time.

File: app/static/joke.js (hunk 17 above)
Comment (fandomario) [reply app/static/joke.js]: Now it is not :)
I checked.

File: app/internal/load_jokes.py (hunk 16 above)
Comment (fandomario) [reply app/internal/load_jokes.py]: I changed that

File: app/main.py (hunk 18 above)
Comment (fandomario) [reply app/main.py]: I have changed that.
Now I am using json_data_loader

**Review by yammesicka — APPROVED** [review]

**Pull request merged**